gcs/object.py

```diff
@@ -495,6 +495,20 @@ def streamer():
         headers["x-goog-metageneration"] = self.metadata.metageneration
         headers["x-goog-storage-class"] = self.metadata.storage_class
 
+        if self.metadata.content_type:
+            headers["Content-Type"] = self.metadata.content_type
```

Comment:
Can you remind me how this interacts with decompressive transcoding?  What is supposed to go here when the data is decompressed?  What about when it is not?

Reply:
My understanding (primarily from the standpoint of http, less storage specific), is outlined in the following table:

| object metadata                                       | request                                 | expected response                                                                                                                                                                                |
|-------------------------------------------------------|-----------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `{contentType:"text/plain"}`                          | `Accept: *`                             | `Content-Type: text/plain`<br/> + raw bytes, without any transform                                                                                                                               |
| `{contentType:"text/plain", contentEncoding: "gzip"}` | `Accept: *`                             | `Content-Type: text/plain`<br/> + server decompressed bytes                                                                                                                                      |
| `{contentType:"text/plain", contentEncoding: "gzip"}` | `Accept: *`<br/>`Accept-Encoding: gzip` | `Content-Type: text/plain`<br/>`Content-Encoding: gzip`<br/> + raw bytes, without any transform                                                                                                  |
| `{contentEncoding: "gzip"}`                           | `Accept: *`<br/>`Accept-Encoding: gzip` | `Content-Type: application/octet-stream`<br/>`Content-Encoding: gzip`<br/> + raw bytes, without any transform                                                                                    |
| `{}`                                                  | `Accept: *`<br/>`Accept-Encoding: gzip` | `Content-Type: application/octet-stream`<br/> + raw bytes, without any transform<br/>OR<br/>`Content-Type: application/octet-stream`<br/>`Content-Encoding: gzip`<br/> + server compressed bytes |
| `{}`                                                  | `Accept: *`                             | `Content-Type: application/octet-stream`<br/> + raw bytes, without any transform                                                                                                                 |

Reply:
Thanks.

Reply:
For posterity (courtesy of @cojenco) https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding dives into the header expectations from GCS's standpoint when Decompressive transcoding is applied.

She also helped me better understand the various pre-computation at play in rest_media such that we can selectively include `Content-Encoding` only when Decompressive transcoding is not used.